056-zk-kraft-migration.md

```diff
@@ -0,0 +1,258 @@
+Starting from Apache Kafka 4.0 (April 2024) only KRaft mode will be supported.
+
+Based on the above Apache Kafka community plan, more details in [KIP-833](https://cwiki.apache.org/confluence/display/KAFKA/KIP-833%3A+Mark+KRaft+as+Production+Ready), the Strimzi project needs to be prepared to allow users to migrate their current clusters from running in ZooKeeper mode to KRaft mode.
+
+## Proposal
+
+The ZooKeeper to KRaft migration implementation is defined by [KIP-866](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration).
+The upstream Apache Kafka community documentation also provides a [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) guide showing how to run the procedure manually on your cluster.
+
+One of the main points to highlight is that such a process doesn't support migration to a KRaft-based cluster running in "combined" mode:
+
+* Cannot migrate from a ZooKeeper based configuration to a cluster with nodes hosting both controller and broker (so actually reducing the number of nodes by removing the ZooKeeper ones).
+* The migration works by having ZooKeeper nodes to be replaced by new KRaft controller nodes.
+* Supporting "combined" mode is actually one of the rejected [alternatives](https://cwiki.apache.org/confluence/display/KAFKA/KIP-866+ZooKeeper+to+KRaft+Migration#KIP866ZooKeepertoKRaftMigration-CombinedModeMigrationSupport) in KIP-866.
+
+It means that the user environment should be sized to host more nodes running in parallel, because during the migration process both ZooKeeper nodes and KRaft controllers will be running.
+The end of the migration will involve shutting down the ZooKeeper nodes reducing the resources consumption.
+
+### Prerequisites
+
+An Apache Kafka cluster is running in ZooKeeper mode with `N` brokers and `M` ZooKeeper nodes, as defined by the `replicas` fields in the corresponding `Kafka` custom resource.
+The cluster has to be migrated to use `KafkaNodePool`(s) first for running the brokers (ZooKeeper nodes are never part of pools).
+Without them it’s not possible to deploy the new KRaft controllers (separated from the current Kafka brokers) as the first step of the migration.
+So the prerequisite is to have the Kafka cluster still using ZooKeeper but running the Kafka brokers with `KafkaNodePool`(s).
+The user can follow the procedure described [here](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str) in the official documentation, in order to meet this prerequisite.
+
+Before migrating to KRaft, the cluster must be upgraded to Kafka version minimum of 3.5.0 with `inter.broker.protocol.version` set to “3.5”.
+
+Furthermore, the running Apache Kafka cluster should not be using JBOD because in this case the migration to KRaft is not supported.
```

Comment:
Should this be really covered here? I would expect that this has to be addressed in Kafka and should not impact the actual migration.

Reply:
But do we know if it will be addressed in Kafka before we have the migration process out?
I mean if we are faster, we should have a check to "block" the migration process because JBOD is still not supported otherwise we still have blockers for starting to work on migration when KRaft in Kafka is not ready.
Not starting the work now would be running faster later.

Reply:
Honestly, I don't know. Kafka is a sh*tshow lately with their post-release deprecations etc. But I think we should focus on how it will look when it is _production-ready_. Also, if you assume it will not be done, do we even want to support KRaft without it? What will we do with the existing users?

Reply:
> if you assume it will not be done

I didn't say that, or I didn't mean that if you got this way. I assume they will support everything in KRaft at some point but should we wait until that point to start working on the migration? It would be maybe late and we cannot promise to the community to have the migration in no time. It's not trivial, it's a complex task, it will take time to develop and test imho. I would just like trying to start as soon as possible.
I could even not adding the check on JBOD making just the assumption that it works in Kafka ... but what if it doesn't when we are ready with the migration but Kafka is not ready yet (i.e. JBOD fixed)?

Reply:
Well, as I said, I think you should plan for the final state. I do not think JBOD itself has impact on how the migration will work given the controller nodes will likely not use it anyway, or? So I would simply not cover it in the proposal.

Reply:
I can remove this sentence. @fvaleri wdyt? You raised the same concern in other comments IIRC.

Reply:
I raised this because JBOD support doesn't seem to be a priority. The KIP is not yet approved. That said, a KRaft only release without JBOD support is unlikely IMO, so we may treat this as an implementation detail, and remove that phrase.